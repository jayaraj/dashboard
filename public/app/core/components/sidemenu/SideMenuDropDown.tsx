import React from 'react';
import DropDownChild from './DropDownChild';
import { NavModelItem } from '@grafana/data';
import { IconName, Link } from '@grafana/ui';
import { css, cx } from '@emotion/css';

interface Props {
  headerText: string;
  headerUrl?: string;
  items?: NavModelItem[];
  onHeaderClick?: () => void;
  reverseDirection?: boolean;
  subtitleText?: string;
}

const SideMenuDropDown = ({
  headerText,
  headerUrl,
  items = [],
  onHeaderClick,
  reverseDirection = false,
  subtitleText,
}: Props) => {
  const resetButtonStyles = css`
    background-color: transparent;
    font-size: inherit;
  `;
  const headerContent = <span className="sidemenu-item-text">{headerText}</span>;
  const header = headerUrl ? (
    <Link href={headerUrl} onClick={onHeaderClick} className="side-menu-header-link">
      {headerContent}
    </Link>
  ) : (
    <button onClick={onHeaderClick} className={cx(resetButtonStyles, 'side-menu-header-link')}>
      {headerContent}
    </button>
  );

  const menuClass = css`
    flex-direction: ${reverseDirection ? 'column-reverse' : 'column'};
  `;

  return (
    <ul className={`${menuClass} dropdown-menu dropdown-menu--sidemenu`} role="menu">
      <li className="side-menu-header">{header}</li>
      {items
        .filter((item) => !item.hideFromMenu)
        .map((child, index) => (
          <DropDownChild
            key={`${child.url}-${index}`}
            isDivider={child.divider}
            icon={child.icon as IconName}
            onClick={child.onClick}
            target={child.target}
            text={child.text}
            url={child.url}
          />
        ))}
      {subtitleText && (
        <li className="sidemenu-subtitle">
          <span className="sidemenu-item-text">{subtitleText}</span>
        </li>
      )}
    </ul>
  );
};

export default SideMenuDropDown;
