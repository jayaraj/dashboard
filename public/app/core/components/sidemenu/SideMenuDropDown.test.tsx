import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter } from 'react-router-dom';
import SideMenuDropDown from './SideMenuDropDown';

describe('SideMenuDropDown', () => {
  const mockHeaderText = 'MyHeaderText';
  const mockHeaderUrl = '/route';
  const mockOnHeaderClick = jest.fn();
  const mockItems = [
    {
      text: 'First link',
    },
    {
      text: 'Second link',
    },
  ];

  it('displays the header text', () => {
    render(<SideMenuDropDown headerText={mockHeaderText} />);
    const text = screen.getByText(mockHeaderText);
    expect(text).toBeInTheDocument();
  });

  it('attaches the header url to the header text if provided', () => {
    render(
      <BrowserRouter>
        <SideMenuDropDown headerText={mockHeaderText} headerUrl={mockHeaderUrl} />
      </BrowserRouter>
    );
    const link = screen.getByRole('link', { name: mockHeaderText });
    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute('href', mockHeaderUrl);
  });

  it('calls the onHeaderClick function when the header is clicked', () => {
    render(<SideMenuDropDown headerText={mockHeaderText} onHeaderClick={mockOnHeaderClick} />);
    const text = screen.getByText(mockHeaderText);
    expect(text).toBeInTheDocument();
    userEvent.click(text);
    expect(mockOnHeaderClick).toHaveBeenCalled();
  });

  it('displays the items', () => {
    render(<SideMenuDropDown headerText={mockHeaderText} items={mockItems} />);
    mockItems.forEach(({ text }) => {
      const childItem = screen.getByText(text);
      expect(childItem).toBeInTheDocument();
    });
  });
});
