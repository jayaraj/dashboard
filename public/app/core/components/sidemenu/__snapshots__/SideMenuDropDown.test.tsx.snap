// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render should not render hideFromMenu children 1`] = `
<ul
  className="dropdown-menu dropdown-menu--sidemenu"
  role="menu"
>
  <li
    className="side-menu-header"
  >
    <a
      className="side-menu-header-link"
    >
      <span
        className="sidemenu-item-text"
      >
        link
      </span>
    </a>
  </li>
  <DropDownChild
    key="undefined-0"
  />
  <DropDownChild
    key="undefined-1"
  />
</ul>
`;

exports[`Render should render children 1`] = `
<ul
  className="dropdown-menu dropdown-menu--sidemenu"
  role="menu"
>
  <li
    className="side-menu-header"
  >
    <a
      className="side-menu-header-link"
    >
      <span
        className="sidemenu-item-text"
      >
        link
      </span>
    </a>
  </li>
  <DropDownChild
    key="undefined-0"
  />
  <DropDownChild
    key="undefined-1"
  />
  <DropDownChild
    key="undefined-2"
  />
</ul>
`;

exports[`Render should render component 1`] = `
<ul
  className="dropdown-menu dropdown-menu--sidemenu"
  role="menu"
>
  <li
    className="side-menu-header"
  >
    <a
      className="side-menu-header-link"
    >
      <span
        className="sidemenu-item-text"
      >
        link
      </span>
    </a>
  </li>
</ul>
`;
